clib/feldspar_array.h

```diff
@@ -65,63 +62,56 @@ static inline struct array* allocArray( struct array* src )
   return src;
 }
 
-/// Resizing an existing array (the struct array must be initialized).
-static inline struct array* resizeArray( struct array* arr, int32_t size, int32_t len )
+/// Resizing an existing array.
+static inline void* resizeArray( void* arr, int32_t size, int32_t len )
 {
-  assert( arr && "array not initialized" );
-  arr->buffer = realloc( arr->buffer, len*size );
-  arr->length = len;
-  return arr;
+  log_3( "resize %p with size %d and len %d\n", arr, size, len );
+  return realloc( arr, len*size );
 }
 
 /// Array (re)initialization for flat arrays.
-static inline struct array* initArray( struct array* src, int32_t size, int32_t len )
+static inline void* initArray( void* arr, int32_t arrLen, int32_t size, int32_t newLen )
 {
-  src = allocArray( src );
-  if( len != src->length ) {
-    src = resizeArray( src, size, len );
+  log_4( "initArray %p with arrlen %d size %d newLen %d\n", arr, arrLen, size, newLen );
+  if( newLen != arrLen ) {
+    arr = resizeArray( arr, size, newLen );
   }
-  return src;
+  return arr;
 }
 
 /// Free a flat array or an array where all the arrays it contains have been free'd already.
 // TODO: Think about arrays escaping from their scope.
-static inline void freeArray( struct array* arr )
+static inline void freeArray( void* arr, int32_t len )
 {
+  log_2( "freeArray %p with length %d\n", arr, len );
   if( arr != NULL ) {
```

Comment:
Null check was required for the dereference, it's not needed for calling free.

Reply:
Right.